### PR TITLE
Cosipbft: recover from a block propagation failure

### PR DESCRIPTION
**Pull request by Gilthoniel**
This improves the round procedure to handle cases where a block is
committed but the leader failed to propagate it to followers.

### CODE REVIEW

**Comment by Gilthoniel** [comment]
The address implementation of Minogrpc had to move to the session package because of cycling imports.

File: core/ordering/cosipbft/pbft/mod.go [hunk 1]
```diff
@@ -186,33 +188,55 @@ func (m *pbftsm) GetViews() map[mino.Address]View {
...
+	if uint16(index) != m.round.leader {
+		return id, xerrors.Errorf("'%v' is not the leader", from)
+	}
+
+	// Check the state after verifying that the proposal comes from the right
+	// leader.
+	if m.state == PrepareState || m.state == CommitState {
+		// The leader should only propose one block, therefore the accepted
+		// proposal identifier is sent back, whatever the input is.
+		return id, nil
```
Comment (nkcr) [inline core/ordering/cosipbft/pbft/mod.go:232]: then what happens to the block? The leader will wait for this block to be committed, which will reach a timeout?

File: core/ordering/cosipbft/pbft/mod.go [hunk 2]
```diff
@@ -226,12 +250,7 @@ func (m *pbftsm) Commit(id types.Digest, sig crypto.Signature) error {
...
 	defer m.Unlock()
 
-	if m.state == CommitState {
-		// The state machine is already committed to a proposal.
-		return nil
-	}
-
-	if m.state != PrepareState {
+	if m.state != PrepareState && m.state != CommitState {
 		return xerrors.Errorf("mismatch state %v != %v", m.state, PrepareState)
```
Comment (nkcr) [inline core/ordering/cosipbft/pbft/mod.go:254]: need an error message update

File: mino/minoch/mod.go [hunk 3]
```diff
@@ -12,15 +12,20 @@ import (
 	"go.dedis.ch/dela/serde/json"
 )
 
+// Filter is a function called for any request to an RPC which will drop it if
+// it returns false.
+type Filter func(mino.Request) bool
```
Comment (nkcr) [inline mino/minoch/mod.go:17]: isn't it a bit expensive to add all this mechanism just for the sake of testing ?

File: mino/minogrpc/session/addr.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+// NewAddress creates a new address.
+func NewAddress(host string) Address {
+	return Address{host: host}
+}
+
+// GetDialAddress returns a string formatted to be understood by grpc.Dial()
+// functions.
+func (a Address) GetDialAddress() string {
+	return a.host
```
Comment (nkcr) [inline mino/minogrpc/session/addr.go:46]: we actually never check that a.host is valid

File: mino/minogrpc/session/addr.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+// FromText implements mino.AddressFactory. It returns an instance of an
+// address from a byte slice.
+func (f AddressFactory) FromText(text []byte) mino.Address {
+	if len(text) == 0 {
+		return Address{}
+	}
+
+	if string(text)[0] == orchestratorCode[0] {
+		return Address{host: string(text)[1:], orchestrator: true}
```
Comment (nkcr) [inline mino/minogrpc/session/addr.go:141]: it panics if len(text) == 1

File: mino/minogrpc/session/addr.go (hunk 4 above)
Comment (Gilthoniel) [reply mino/minogrpc/session/addr.go:46]: Well if that's the case, it will fail to dial.

File: mino/minoch/mod.go (hunk 3 above)
Comment (Gilthoniel) [reply mino/minoch/mod.go:17]: Minoch has been implemented for the sake of testing, so I don't think so. It's also very cheap as the default behaviour of being empty does add anything to the logic.

File: core/ordering/cosipbft/pbft/mod.go (hunk 1 above)
Comment (Gilthoniel) [reply core/ordering/cosipbft/pbft/mod.go:232]: Yes, if the leader decides to propose a different block, the followers will simply reject it and proceed to a view change after the timeout. Then new leader == new block.

File: mino/minogrpc/session/addr.go (hunk 5 above)
Comment (Gilthoniel) [reply mino/minogrpc/session/addr.go:141]: it actually doesn't: https://play.golang.org/p/LYMPrve_bQK

`[2:]` would panic though.

Comment (Gilthoniel) [reply mino/minogrpc/session/addr.go:141]: I'll add a test anyway.

**Review by nkcr — APPROVED** [review]

**Comment by nkcr** [comment]
Can be merged once the conflict is resolved.

**Comment by Gilthoniel** [comment]
I'll wait for #125 to be merged, otherwise it will be pain to rebase it.

**Pull request merged**